### ISSUE TITLE
check_http [-e] is very fragile/broken after "-f fix"

### ISSUE THREAD

**Issue by cejkar** [nagios-plugins/nagios-plugins#243]
- If there is missing any empty line in the response or the response is empty (which is something we have to test, atleast empty response), check_http now ends in active endless loop terminated by connection timeout. Many embedded web servers do not return any header line and start immediately with &lt;html&gt;, web browsers do not have any problem with this and previous version 2.1.1 was usable with -e "&lt;html&gt;". Solution would be to change for (;;) to while (*page != '\0'), but still for web servers without headers we have to use another tool now.

- If there is missing space in the status line, check_http dumps core now (like "&lt;html&gt;" as a first response line). strchr (status_line, ' ') has to be tested on NULL return value, for example: space_code = strchr (status_line, ' '); status_code = space_code + sizeof (char); if (space_code == NULL || strspn (status_code, "1234567890") != 3) ...

- Statement "If (-e) specified skips all other status line logic (ex: 3xx, 4xx, 5xx processing)" is no longer true and for somebody -e is broken now. Previous version returned OK state, current version for example returns HTTP WARNING: Status line output matched "401" - HTTP/1.1 401 Unauthorized. And please note that redirects are 3xx status codes, so according to -e specification should be skipped, which means that previous behavior was correct and -f should not be used with -e, because these two options are conflicting. And solution would be to revert "-f fix", or I have used if (!server_expect_yn) { ... } around status line logic.

**Comment by cejkar** [comment]
Second try: GitHub eats &lt; and &gt; tags...

**Comment by cejkar** [comment]
Yes, ok, so once more:

- If there is missing any empty line in the response or the response is empty (which is something we have to test, atleast empty response), check_http now ends in active endless loop terminated by connection timeout. Many embedded web servers do not return any header line and start immediately with &lt;html&gt;, web browsers do not have any problem with this and previous version 2.1.1 was usable with -e "&lt;html&gt;". Solution would be to change for (;;) to while (*page != '\0'), but still for web servers without headers we have to use another tool now.

- If there is missing space in the status line, check_http dumps core now (like "&lt;html&gt;" as a first response line). strchr (status_line, ' ') have to be tested on NULL return value, for example: space_code = strchr (status_line, ' '); status_code = space_code + sizeof (char); if (space_code == NULL || strspn (status_code, "1234567890") != 3) ...

- Statement "If (-e) specified skips all other status line logic (ex: 3xx, 4xx, 5xx processing)" is no longer true and for somebody -e is broken now. Previous version returned OK state, current version for example returns HTTP WARNING: Status line output matched "401" - HTTP/1.1 401 Unauthorized. And please note that redirects are 3xx status codes, so according to -e specification should be skipped, which means that previous behavior was correct and -f should not be used with -e, because these two options are conflicting. And solution would be to revert "-f fix".

**Comment by yanickgirouard** [comment]
I would expect the -e switch to check the status of the final redirected page rather than the first one it finds when combined with the -f switch. It makes no sense to stop at the first response if the -f switch is used because the whole point of using the -f switch is to test against a redirected page and not the page that generates the redirect. The issue is that the -e check is done before the redir function calls. The whole logic should be reviewed.

**Comment by jfrickson** [comment]
Your first two issues should be pretty easy. Looking into them now.

For your third issue, there was a lot of discussion and the latest version mostly satisfied everybody. I'll take another look at it, though.

**Comment by sshmanko** [comment]
About the third point, -e argument should ignore HTTP status code, otherwise it is unusable for non 2xx responses:

./check_http -H getstatuscode.com -u /403 --ssl -e 'HTTP'

v2.1.1:
`HTTP OK: Status line output matched "HTTP" - 37541 bytes in 7.397 second response time |time=7.396698s;;;0.000000 size=37541B;;;0`

v.2.2.0:
`HTTP WARNING: Status line output matched "HTTP" - HTTP/1.1 403 Forbidden - 37610 bytes in 8.202 second response time |time=8.202089s;;;0.000000 size=37610B;;;0`

**Comment by pbiering** [comment]
there is a fix available somewhere, see also: https://bugzilla.redhat.com/show_bug.cgi?id=1426817

**Comment by syl-robitaille** [comment]
Unfortunately the link provided earlier to RedHat's Bugzilla entry
doesn't provide the patch they've applied.  With respect to the third
issue described above, I found this problem report while searching in
preparation to reporting exactly this issue.  The RedHat fix refers
to a logic error, and that is exactly what I believe to be the cause,
and what I was going to report.  I've applied the following fix to
our local installation of Nagios-Plugins-2.2.0, and it is so far
functioning as intended:

    --- ./plugins/check_http.c	2017-01-19 11:01:31.000000000 -0500
    +++ ./plugins/check_http.c	2017-04-05 14:59:26.617387861 -0400
    @@ -1220,6 +1220,9 @@
         printf ("**** HEADER ****\n%s\n**** CONTENT ****\n%s\n", header,
                     (no_body ? "  [[ skipped ]]" : page));
     
    +  /* NOTE: After the following blocks, msg *MUST* be an asprintf-allocated string */
    +  xasprintf (&msg, "");
    +
       /* make sure the status line matches the response we are looking for */
       if (!expected_statuscode (status_line, server_expect)) {
         if (server_port == HTTP_PORT)
    @@ -1234,16 +1237,12 @@
       }
     
       /* Bypass normal status line check if server_expect was set by user and not default */
    -  /* NOTE: After this if/else block msg *MUST* be an asprintf-allocated string */
    -  if ( !bad_response ) {
         if ( server_expect_yn  )  {
           xasprintf (&msg,
                     _("Status line output matched \"%s\" - "), server_expect);
           if (verbose)
             printf ("%s\n",msg);
    -    } else
    -      xasprintf (&msg, "");
    -  }
    +  } else {
     
       /* Status-Line = HTTP-Version SP Status-Code SP Reason-Phrase CRLF */
       /* HTTP-Version   = "HTTP" "/" 1*DIGIT "." 1*DIGIT */
    @@ -1283,6 +1282,7 @@
         /* Print OK status anyway */
         xasprintf (&msg, _("%s%s - "), msg, status_line);
       }
    +  } /* end else (server_expect_yn)  */
     
     	free(status_line);
 
Please note that I have not studied the check_http plugin in sufficient
depth to be prepared to claim that this is the correct fix in the
general case, but it does work here, and I'm submitting it in hope
that it could be useful for others.

-- 
Sylvain Robitaille

**Comment by jfrickson** [comment]
Thanks, @syl-robitaille !

**Comment by hedenface** [comment]
@syl-robitaille Are you able to submit a PR with your patch? That way you get the credit :)

**Comment by syl-robitaille** [comment]
On Sun, 18 Jun 2017, Bryan Heden wrote:

> @syl-robitaille Are you able to submit a PR with your patch? That way you get the credit :)

Possibly ... but what's a "PR" (feel free to just point me to
documentation I should read) ...

-- 
----------------------------------------------------------------------
Sylvain Robitaille                               syl@encs.concordia.ca

Systems analyst / AITS                            Concordia University
Faculty of Engineering and Computer Science   Montreal, Quebec, Canada
----------------------------------------------------------------------

**Comment by hedenface** [comment]
A PR is a GitHub Pull Request. Ours are here: https://github.com/nagios-plugins/nagios-plugins/pulls

You can find the starter documentation here: https://help.github.com/articles/about-pull-requests/

But there is plenty of other documentation regarding pull requests available via official GitHub channels as well :)

Good luck!

**Comment by syl-robitaille** [comment]
On Mon, 19 Jun 2017, Bryan Heden wrote:

> A PR is a GitHub Pull Request. ...

Excellent.  Thanks.  I'll read the docs and do what I need to do.

-- 
----------------------------------------------------------------------
Sylvain Robitaille                               syl@encs.concordia.ca

Systems analyst / AITS                            Concordia University
Faculty of Engineering and Computer Science   Montreal, Quebec, Canada
----------------------------------------------------------------------

**Comment by syl-robitaille** [comment]
On Mon, 19 Jun 2017, Bryan Heden wrote:

> A PR is a GitHub Pull Request. ...

With MUCH thanks for the links to information!

I believe I've now dones what needs to be done, and submitted my
first ever GIT pull request.  ... and I'm reasonably sure I didn't
break anything on the real branch!  ;-)

-- 
----------------------------------------------------------------------
Sylvain Robitaille                               syl@encs.concordia.ca

Systems analyst / AITS                            Concordia University
Faculty of Engineering and Computer Science   Montreal, Quebec, Canada
----------------------------------------------------------------------

**Comment by hedenface** [comment]
@syl-robitaille I ran some styling utility and it messed up the ability to cleanly merge your PR. You're added to the THANKS file in the upcoming commit.

I took a lot of what everyone here said into consideration and tried my best - but check_http is kind of a beast. I have big plans for plugins3 - I think we're going to break this into the separate checks it needs to be.

Regardless, I believe I have a fix here: https://github.com/nagios-plugins/nagios-plugins/commit/2bc8d5e7499a9afea1fda9af0b06dc2b58e5f8dd

Also - `-e` functionality will stay as is. I agree that it shouldn't check the status against the redirected page. If you're checking a status, it makes sense to check the status of the host/site/page/whatever you're specifying.
